sensor/debugger/k8s/fake.go

```diff
@@ -0,0 +1,296 @@
+		},
+		serviceKind:    func(namespace string) interface{} { return f.Client.Kubernetes().CoreV1().Services(namespace) },
+		jobKind:        func(namespace string) interface{} { return f.Client.Kubernetes().BatchV1().Jobs(namespace) },
+		replicaSetKind: func(namespace string) interface{} { return f.Client.Kubernetes().AppsV1().ReplicaSets(namespace) },
+		replicationControllerKind: func(namespace string) interface{} {
+			return f.Client.Kubernetes().CoreV1().ReplicationControllers(namespace)
+		},
+		daemonSetKind:   func(namespace string) interface{} { return f.Client.Kubernetes().AppsV1().DaemonSets(namespace) },
+		deploymentKind:  func(namespace string) interface{} { return f.Client.Kubernetes().AppsV1().Deployments(namespace) },
+		statefulSetKind: func(namespace string) interface{} { return f.Client.Kubernetes().AppsV1().StatefulSets(namespace) },
+		cronJobKind:     func(namespace string) interface{} { return f.Client.Kubernetes().BatchV1().CronJobs(namespace) },
+		podKind:         func(namespace string) interface{} { return f.Client.Kubernetes().CoreV1().Pods(namespace) },
+	}
+	f.resourceMap = map[string]interface{}{
+		namespaceKind:             &corev1.Namespace{},
+		secretKind:                &corev1.Secret{},
+		serviceAccountsKind:       &corev1.ServiceAccount{},
+		roleKind:                  &rbacv1.Role{},
+		clusterRoleKind:           &rbacv1.ClusterRole{},
+		roleBindingKind:           &rbacv1.RoleBinding{},
+		clusterRoleBindingKind:    &rbacv1.ClusterRoleBinding{},
+		networkPolicyKind:         &networkingv1.NetworkPolicy{},
+		nodeKind:                  &corev1.Node{},
+		serviceKind:               &corev1.Service{},
+		jobKind:                   &batchv1.Job{},
+		replicaSetKind:            &appsv1.ReplicaSet{},
+		replicationControllerKind: &corev1.ReplicationController{},
+		daemonSetKind:             &appsv1.DaemonSet{},
+		deploymentKind:            &appsv1.Deployment{},
+		statefulSetKind:           &appsv1.StatefulSet{},
+		cronJobKind:               &batchv1.CronJob{},
+		podKind:                   &corev1.Pod{},
+	}
+}
+
+// WaitForDone waits until all the events have been processed
+func (f *FakeEventsManager) WaitForDone() {
+	doneC := f.done.WaitC()
+	<-doneC
+}
+
+// waitForMinimumResources waits for a minimum number of resources to be created or once all the events have been processed
+func waitForMinimumResources(ch <-chan string, done concurrency.Signal) error {
+	count := 0
+	doneC := done.WaitC()
+	d := false
+	for {
+		select {
+		case obj, more := <-ch:
+			if !more && d {
+				return errors.New("the events file did not contain the minimum resources required to start sensor")
+			}
+			if _, ok := minimumResources[obj]; ok {
+				minimumResources[obj]--
+				if minimumResources[obj] == 0 {
+					count++
+				}
+				if len(minimumResources) == count {
+					return nil
+				}
+			}
+		case <-doneC:
+			// We make sure there are not more messages in the channel
+			obj, more := <-ch
```

Comment:
I think this might lead to a race condition. 

If you're using a buffered channel, it means that the emitter of events (goroutine under `CreateEvents`) is allowed to write in a non-blocking way. Therefore, it could be that once the done signal is called, `ch` has more than one record waiting in the queue. So if we're only reading one here is not sufficient. You'd still have to loop until the channel is empty (or closed).

Reply:
I think it might make more sense to control the flow by simply iterating on the channel until closed rather than using two constructs (a channel and a signal).

Reply:
You're right, it could lead to a race condition. We changed the code in this part too after the chat.

Reply:
Ok. Let me know when it's ready to reviewed again.